### DISCUSSION TITLE
Grott on ESP - A fork idea

### DISCUSSION THREAD

**Discussion by the-butterfry** [johanmeijer/grott#114] [Ideas]
Hey @johanmeijer  I want to say I've been using Grott for a year now and it's all going well enough. Your code just chugs along nicely without drama. I really appreciate that! I have more issues with the Wifi stick than anything.  My Growatt died over the winter and had to get it replaced - Getting my old Wifi stick unregistered and setup on the new unit was really annoying.

On that topic, and since the Wifi-(s/f/x) is just an ESP8266, I went hunting for a fix to see what others may have developed in the last year such as:

- Custom firmware that re-writes the Wifi stick. Their code which only has the http interface on some units and only while in config mode is... obnoxious. The growatt/shinephone server is dodgy. 
- Since it's just an ESP with a RTC, possibly a fully custom ESP with a USB A interface (and maybe a RTC for fun too). Off the shelf would be best since the sticks aren't THAT expensive. By the time you get a custom board made and all the effort, why not flash the original stick?

I found this: https://github.com/otti/Growatt_ShineWiFi-S and started a [conversation ](https://github.com/otti/Growatt_ShineWiFi-S/issues/11#issue-1170519839)about Grott on an ESP. 

**I don't need to do much explaining but a combined project would do a few things for us:**

- Complexity - Eliminate the need for complex upgrades/docker containers/config files. It could be a .bin uploaded with settings saved and web configuration. If you lose your config no big deal - just reflash or upload a saved config. 
- In many cases, eliminate the need for onsite computer to run Grott. If we can send out via MQTT then our data warehouse can subscribe to it.

**Downsides:**

- Eliminating an onsite system computer would be a little limiting as for other system functionality. For example, I am running another program to pull in BMS info that would need a computer anyhow (until I put in the time to run it on my OpenWRT router). 
- If Grott in full can't fit in one bin file, then builds for each inverter type? There's some complexity there. I haven't investigated the size we're looking at. But a modern ESP32 should have enough flash. I guess the other issue might be horsepower and scalability as the number of paralleled units grow. Would we be dealing with a Main Grott stick and a Set of secondaries that report to it? Lots of architectural questions actually... maybe just feasibility for now?
- With a custom build on a generic ESP we'd have Serial # issues preventing us registering/sending data to Growatt. But with a new firmware on a real Wifi stick and a valid Serial #, no problem. 

What do you think? Possible? I like the compactness of the idea.

**Comment by johanmeijer** [comment]
Hi @the-butterfry,

Yes of course this is possible but see it as a separate project. It will not be a Fork of Grott. Grott is written in Python and specific for intercept and interpret the data sent by the Growatt Datalogger (Shiniwifi, shinelan Lanbox etc). 

What you want is to create your own datalogger that communicates (retrieve data from) with the inverter. And yes the Growatt shinewifi is based on an ESP and can be reprogrammed. And yes it is absolute possible. 

It is one of things on my to do list. I have a lot of ESP's (8266, 32) already in use for my home automation. Only programming an ESP (microcontroller) is a little bit different than writing an python program it mostly done in C++ and with an special IDE like arduino or platformio. 

But we can start a project to do it. I am not sure I want to "offer" my current shinewifi for this while I need this also for Grott development and support. But I have enough spare ESP boards and also RS232/Modbus boards "in stock" to make my own"stick".  

An other thing that we can do in stead of creating a new datalogger with an ESP, is creating an Grott instance (Server included) on an ESP.  Then we do not need to reprogram ("offer") the shine sticks. But you also do need a separate device to run Grott. (But be aware you always will need device with applications to receive data (like MQTT, InluxDB, HomeAssistent).  

Be aware it is not only creating a simple program that reads and sends the inverter settings/registers but you also need an interface (HTML?) to configure and manage the logger and also define the interfaces (MQTT, INFLUXDB, PVOutput but als a NTP interface for datetime). Next to it we need an OTA (over the air) possible to reload the programs to eliminate the need of connecting the datalogger to a PC.

**Reply by the-butterfry** [reply]
Yup, I hear that. This is why I got excited about what was going on over at https://github.com/otti/Growatt_ShineWiFi-S
They have the esp with a web interface (or shine stick) already going complete with. Bin uploads and  a simple graph of inverter output. configurations are available at initial flash time. I flashed one already and it works great. We just need to wedge grott on there so the esp version of grott can intercept it. Pretty sure it would be trivial to add grottos ini config to the web config page.
